### PR TITLE
Ticket/15305 streams

### PR DESCRIPTION
**Pull request by rubencm**
Checklist:

- [ ] Correct branch: master for new features; 3.2.x, 3.1.x for fixes
- [ ] Tests pass
- [ ] Code follows coding guidelines: [master / 3.2.x](https://area51.phpbb.com/docs/master/coding-guidelines.html), [3.1.x](https://area51.phpbb.com/docs/31x/coding-guidelines.html)
- [ ] Commit follows commit message [format](https://wiki.phpbb.com/Git#Commit_Messages)

Tracker ticket (set the ticket ID to **your ticket ID**):

https://tracker.phpbb.com/browse/PHPBB3-12345

### CODE REVIEW

File: phpBB/phpbb/storage/adapter/local.php [hunk 1]
```diff
@@ -195,4 +195,39 @@ protected function ensure_directory_exists($path)
 			$this->create_dir($path);
 		}
 	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function read_stream($path)
+	{
+		$stream = @fopen($path, 'rb');
```
Comment (Nicofuma) [inline phpBB/phpbb/storage/adapter/local.php]: I think $this->root_path is missing

File: phpBB/phpbb/storage/adapter/local.php [hunk 2]
```diff
@@ -195,4 +195,39 @@ protected function ensure_directory_exists($path)
...
+
+		return $stream;
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function write_stream($path, $resource)
+	{
+		if ($this->exists($path))
```
Comment (Nicofuma) [inline phpBB/phpbb/storage/adapter/local.php:220]: I think $this->root_path is missing

Comment (rubencm) [reply phpBB/phpbb/storage/adapter/local.php:220]: No, it is inside exists method

File: phpBB/phpbb/storage/adapter/local.php (hunk 1 above)
Comment (rubencm) [reply phpBB/phpbb/storage/adapter/local.php]: yes, thank you

File: phpBB/phpbb/storage/adapter/local.php (hunk 2 above)
Comment (Nicofuma) [reply phpBB/phpbb/storage/adapter/local.php:220]: ok

**Comment by Nicofuma** [comment]
For the streams I think it should stay in another interface (it removes the not supported exception) and then, in the storage, if the adapter does not support streams we could emulate them:
1. Copy the file to a temp directory
2. Stream the temp file

**Comment by rubencm** [comment]
Well, that emulation could be done in the adapters.
About the not_implemented i think i will need it for the get_mimetype and get_filesize methods i have to add in "avatars"

**Comment by Nicofuma** [comment]
Honestly I think it is better to emulate it in the storage than adapter because this way we won't duplicate it.
About get_mimetype and get_filesize, you will go through the storage class, right? So you can deal with it here too

**Comment by rubencm** [comment]
Ok, here in this pr?

**Comment by rubencm** [comment]
I dont know if is a good idea to emulate it, because with avatars maybe ok, but with attachments the script can get out of memory.

**Comment by Nicofuma** [comment]
we will always have the issue unless we have streams or a way to dump the file to the disk without loading it in memory (it's streams to be honest...)

Do you have in mind some cloud providers that does not support streams?

**Comment by rubencm** [comment]
I dont know, all providers should support streams, but probably not all their apis

File: phpBB/phpbb/storage/storage.php [hunk 3]
```diff
@@ -133,11 +140,64 @@ public function rename($path_orig, $path_dest)
...
+	 *										When target file cannot be created
+	 * @throws \phpbb\storage\exception\not_implemented	When the adapter doesnt implement the method
+	 */
+	public function write_stream($path, $resource)
+	{
+		$adapter = $this->get_adapter();
+
+		if ($adapter instanceof stream_interface)
+		{
+			$adapter>write_stream($path, $resource);
```
Comment (rubencm) [inline phpBB/phpbb/storage/storage.php]: Wtf, how was this working?

File: phpBB/phpbb/storage/storage.php [hunk 4]
```diff
@@ -133,11 +140,64 @@ public function rename($path_orig, $path_dest)
...
+		$adapter = $this->get_adapter();
+
+		if ($adapter instanceof stream_interface)
+		{
+			$stream = $adapter->read_stream($path);
+		}
+		else
+		{
+			// Simulate the stream
+			$stream = tmpfile();
```
Comment (rubencm) [inline phpBB/phpbb/storage/storage.php:173]: I am looking that this doesn't create the file in binary mode, does it matter?

Comment (Nicofuma) [reply phpBB/phpbb/storage/storage.php:173]: I think yes, to solve the issue you can combine tempname and fopen

**Pull request closed without merging**